src/coreclr/jit/codegenarmarch.cpp

```diff
@@ -815,32 +806,35 @@ void CodeGen::genPutArgStk(GenTreePutArgStk* treeNode)
             else
 #endif // TARGET_ARM64
             {
-                emitAttr storeAttr = emitTypeSize(targetType);
+                emitAttr storeAttr = emitTypeSize(source->TypeGet());
                 emit->emitIns_S_R(INS_str, storeAttr, srcReg, varNumOut, argOffsetOut);
                 argOffsetOut += EA_SIZE_IN_BYTES(storeAttr);
             }
             assert(argOffsetOut <= argOffsetMax); // We can't write beyond the outgoing arg area
             return;
         }
 
+        var_types slotType = genActualType(source);
         if (compMacOsArm64Abi())
```

Comment:
@jakobbotsch, just for my own education; is it specific to mac ABI? i.e. if we apply the same classification to linux and windows arm64, it will break putarg on those targets?

This TODO is probably related: https://github.com/dotnet/runtime/blob/d682d680684773530cfcf1c5fc7650dd640ddda9/src/coreclr/jit/morph.cpp#L736-L738

Reply:
macOS arm64 ABI is different in that multiple small arguments can be packed into the same stack slot. It means we have different behavior in some places in the JIT, and that different behavior was wrong here.
The behavior for the other ABIs did not have this bug.